compiler/rustc_hir_typeck/src/lib.rs

```diff
@@ -236,9 +236,17 @@ fn infer_type_if_missing<'tcx>(fcx: &FnCtxt<'_, 'tcx>, node: Node<'tcx>) -> Opti
             && let ty::AssocKind::Const = item.kind
             && let ty::ImplContainer = item.container
             && let Some(trait_item) = item.trait_item_def_id
+            // We don't want to call `tcx.compare_impl_const` here to avoid cycle errors. This is a
+            // very naive check to prevent us from messing up the generic param instantiation below.
+            && tcx.generics_of(def_id).own_counts() == tcx.generics_of(trait_item).own_counts()
```

Comment:
Can we use https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/context/struct.TyCtxt.html#method.check_args_compatible ? Perhaps after rebasing the const onto the impl?

Reply:
Like, similar to what we do when we have a mismatch between GAT params in a trait and impl

Reply:
Thanks for the pointer, I'll check!

Reply:
relevant callsite: https://github.com/rust-lang/rust/blob/9c8a58fdb895204cb19eeb97472a78caa1c57c19/compiler/rustc_trait_selection/src/traits/project.rs#L2108-L2115

Reply:
Applied. Great idea, thanks!